components/iam/pkg/oidc/service.go

```diff
@@ -157,30 +157,34 @@ func randString(size int) (string, error) {
 	return base64.RawURLEncoding.EncodeToString(b), nil
 }
 
-func (s *Service) GetClientConfigFromRequest(r *http.Request) (*ClientConfig, error) {
+func (s *Service) GetClientConfigFromStartRequest(r *http.Request) (*ClientConfig, error) {
 	issuerParam := r.URL.Query().Get("issuer")
-	stateParam := r.URL.Query().Get("state")
-	if issuerParam == "" && stateParam == "" {
-		return nil, fmt.Errorf("missing request parameters")
+	if issuerParam == "" {
+		return nil, fmt.Errorf("missing issuer parameter")
 	}
 
-	if issuerParam != "" {
-		for _, value := range s.configsById {
-			if value.Issuer == issuerParam {
-				return value, nil
-			}
+	for _, value := range s.configsById {
+		if value.Issuer == issuerParam {
+			return value, nil
 		}
 	}
 
-	if stateParam != "" {
-		state, err := decodeStateParam(stateParam)
-		if err != nil {
-			return nil, fmt.Errorf("bad state param")
-		}
-		config := s.configsById[state.ClientConfigID]
-		if config != nil {
-			return config, nil
-		}
+	return nil, fmt.Errorf("failed to find OIDC config for start request")
+}
+
+func (s *Service) GetClientConfigFromCallbackRequest(r *http.Request) (*ClientConfig, error) {
+	stateParam := r.URL.Query().Get("state")
+	if stateParam == "" {
+		return nil, fmt.Errorf("missing state parameter")
+	}
+
+	state, err := decodeStateParam(stateParam)
+	if err != nil {
+		return nil, fmt.Errorf("bad state param")
+	}
+	config := s.configsById[state.ClientConfigID]
+	if config != nil {
+		return config, nil
 	}
```

Comment:
Why do we need to modify the service state inside a middleware? This causes the middleware to be coupled to the service.

If we need to extract config, we can just call a function to extract it as part of the handler, and not store it. This feels like a premature optimization which makes it hard to follow what is happening as the middleware causes side-effects.

Reply:
It would look like this https://github.com/gitpod-io/gitpod/compare/at/refactor-config-mw-2?expand=1

Reply:
I do not agree on that at all, as it creates bloated handlers, and the argument sounds like it's just a personal taste. 
Anyways, for the sake of getting things done, I'll rewrite just to not loose focus on the more important things.

Reply:
done

Reply:
>, and the argument sounds like it's just a personal taste

It's not personal taste. It's based on the following.
* Explicit is better than implicit. If I can make something very direct, without indirection, it's often easier to read, and hence get it correct, than indirect (`provider` is the indirection here)
* In this example, we're not re-using that middleware for multiple routes. As such, we get better readability by handling it directly, rather than through an indirection - middleware
* I also want to reduce bloat. And for me, bloat are the things that are there, which do not need to be there - concepts, indirections, abstractions. I mentioned this, because for me this felt like an unnecessary indirection.

Thanks for making the change.